tools/afl-fuzzer/main.c

```diff
@@ -0,0 +1,379 @@
+		ret = pthread_cond_timedwait(&ipc_cond, &ipc_mutex, &timeout);
+
+	if (ret == ETIMEDOUT && !ipc_reply_recd) {
+		ret = -EINVAL;
+		fprintf(stderr, "error: IPC timeout\n");
+		ipc_dump_err(&fuzzer->msg);
+		pthread_mutex_unlock(&ipc_mutex);
+		exit(0);
+	}
+
+	pthread_mutex_unlock(&ipc_mutex);
+
+	/*
+	 * sleep for 5 ms before continuing sending the next message.
+	 * This helps with the condition signaling. Without this,
+	 * the condition seems to always satisfy and
+	 * the fuzzer never waits for a response from the DSP.
+	 */
+	usleep(50000);
+
+	return ret;
+}
+
+int main(int argc, char *argv[])
+{
+	struct fuzz fuzzer;
+	int ret;
+	char *ipc_msg_file = NULL;
+	char *platform_name = "byt";
+	char opt;
+	FILE *fp = NULL;
+
+	while ((opt = getopt(argc, argv, "m:")) != -1) {
+		switch (opt) {
+		case 'm':
+			ipc_msg_file = optarg;
+			break;
+		default:
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (argc != 2) {
+		fprintf(stderr, "error: Usage ./sof-afl-fuzzer file\n");
+		exit(EXIT_FAILURE);
+	}
+
+	/* init platform */
+	fprintf(stdout, "initialising platform %s\n", platform[0]->name);
+	ret = platform[0]->init(&fuzzer, platform[0]);
+
+	if (ret == ETIMEDOUT) {
+		fprintf(stderr, "error: platform %s failed to initialise\n",
+			platform_name);
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(stdout, "FW boot complete\n");
+
+	/* allocate max ipc size bytes for the msg and reply */
+	fuzzer.msg.msg_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+	fuzzer.msg.reply_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+
+	/* Open ipc message file */
+	fp = fopen(argv[1], "rb");
+	if (!fp) {
+		fprintf(stderr, "error: opening ipc msg file %s\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* ipc message file structure:
+	   header
+	   msg_size
+	   msg data
+	*/
+
+	ret = fread(&fuzzer.msg.header, sizeof(uint32_t), 1, fp);
+	
+	if (ret != 1) {
+		fprintf(stderr, "error: reading header of ipc message from %s failed\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(stdout, "header read: %ud\n", fuzzer.msg.header);
+
+	unsigned int uret = 0;	
+	uret = fread(&fuzzer.msg.msg_size, sizeof(unsigned int), 1, fp);
+
+	fprintf(stdout, "msg_size read: %lu\n", fuzzer.msg.msg_size);	
+	if (uret != 1) {
+		fprintf(stderr, "error: reading msg_size of ipc message from %s failed\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}
+
+	if (fuzzer.msg.msg_size > SOF_IPC_MSG_MAX_SIZE) {
+		fprintf(stderr, "error: msg_size of ipc message exceeds max size %d\n",
+			SOF_IPC_MSG_MAX_SIZE);
+		exit(EXIT_FAILURE);
+	}		
+
+
+	uret = fread(fuzzer.msg.msg_data, sizeof(char), fuzzer.msg.msg_size, fp);
+
+	if (uret != fuzzer.msg.msg_size) {
+		fprintf(stderr, "error: reading msg_data of ipc message from %s failed\n",
+			argv[1]);
+		exit(EXIT_FAILURE);
+	}		
+	
+	// Initializing fuzzer messsage randomly
+	/* fuzzer.msg.header = SOF_IPC_GLB_COMP_MSG; */
+	/* fuzzer.msg.msg_size = 0; */
+
+	ret = fuzzer_send_msg(&fuzzer);
+
+	/* all done - now free platform */
+	platform[0]->free(&fuzzer);
+	return 0;
```

Comment:
Can you comment each code block here otherwise its difficult to follow what each stage is doing.